add docs for http transport (#95)
smartcontractkit/ea-framework-js: docs/v3-ea-components/transport-types/http-transport.md

# HTTP Transport

HttpTransport is used to fetch data from a Provider using HTTP(S) protocol. In order to use HttpTransport you need to provide two methods `prepareRequests` and `parseResponse` that will be used to build requests and parse responses respectively. 

Example usage of HttpTransport
```typescript  
const transport = new HttpTransport<EndpointTypes>({  
  prepareRequests: (params, config) => {  
    return params.map((param) => {  
      const symbol = param.symbol.toLowerCase()  
      const url = `/price/${symbol}`  
  
      return {  
        params: param  
		request: {  
          baseURL: config.API_ENDPOINT,  
          url,  
        },  
      }  
    })  
  },  
  parseResponse: (params, res, config) => {  
    return res.data.map((result) => {  
      return {  
        params: { symbol: result.symbol},  
        response: {  
          data: {  
            result: result.price,  
          },  
          result: result.price,  
        },  
      }  
    })  
  },  
})
```

> ### Building requests

Reviewer: Another nitpick.

I can't do a suggested change because the formatting gets messed up, but we should have one open line between L36-L37 and L37-L38.

> **prepareRequests**  method  takes the list of currently active input parameters in the subscription set and adapter config and should return one or multiple request configs (_ProviderRequestConfig_) depending if Data Provider supports request batching or not.

Reviewer: Trying to make this a bit more clear:

```suggestion
The **prepareRequests**  method  takes the list of currently active input parameters in the subscription set and the adapter config. If the Data Provider only supports requesting a single data point for each HTTP request, the method should return a single request config (_ProviderRequestConfig_). However, if the Data Provider supports requesting multiple data points in each HTTP request, the method should return an array of request configs.
```

#### Request config structure

> *ProviderRequestConfig* is a object that contains two keys `params` and `request` . `params` is the input parameters for a request and whatever  is provided will be available in the **parseResponse** method. `request` is the *axios* request configuration where request url, request method, query or body params, headers and etc can be specified.

Reviewer: > `params` is the EA input parameters for a request

This^ is accurate, right?

```suggestion
*ProviderRequestConfig* is an object with two keys: `params` and `request`. `params` is the EA input parameters for a request, and will also be available in the **parseResponse** method. `request` is the *axios* request configuration where parameters such as the request URL, request method, query or body params, and headers can be specified.
```

Author: regarding `params` i would say yes and. no :). it will be more accurate to say
 `params` is the non-expired input parameters in the subscription set, but i think it will be more confusing or somewhere we need to also mention what is that subscription set and how it works.

Reviewer: Ok, I'm fine with how it is now!

> #### Send batch request to Data Provider
> The example above assumes that Data Provider doesn't support batching and each active input parameter will initiate request to Data Provider. Bellow is the example of batch version of prepareRequests
> ```typescript  

Reviewer: More spacing

```suggestion
#### Send batch request to Data Provider

The example above assumes that the Data Provider doesn't support batching, and each active input parameter will initiate a request to Data Provider. Below is an example of a batch version of prepareRequests.

```typescript  
```

Reviewer: > The example above assumes

Which example is it referring to?

Author: the very first `Example usage of HttpTransport` .  maybe we can rename it like that ? 'the first example' ?

  prepareRequests: (params, config) => {  
    return {  
        params: params,
		request: {  
          baseURL: config.API_ENDPOINT,  
          url: '/price',
          params: {
	          symbols: [...new Set(params.map((p) => p.symbol.toLowerCase()))].join(',')
          }  
        },  
      } 
  }
```

> The difference is that instead of returning array of request config objects, only one object is returned where information about input parameters is passed to Data Provider as one query parameter.

Reviewer: Spacing

```suggestion

The difference is that instead of returning array of request config objects, only one object is returned where information about input parameters is passed to Data Provider as one query parameter.
```

        },  
      }  
    })  
  }
```
#### Error response structure
It is also possible to store and return errors as well, in case Data Provider has not returned information that was expected.  In case of error, instead of `data` and `result`, the `resonse` object should contain `statusCode` for the errored response and `errorMessage` that will be sent back from the adapter.

Refactored example that shows both error and success examples
```typescript
parseResponse: (params, response, config) => {  
    return params.map((param) => {
	  const value = response.data[param.symbol]?.price  
	  if (!value) {
		  // no value provided from DP. Sending error response back
		  return {
			   params: param,
			   response: {
				   errorMessage: `Could not retrieve valid data from Data Provider for symbol ${param.symbol}`,
				   statusCode: 502
			   }
		  }
	  }
	  // if everything is ok we send success response
      return {  
        params: param, // Note that this should be single input parameter, not the params array above
        response: {  
          data: {  
            result: value,  
          },  
          result: value,  
        },  
      }  
    })  

>   }

Reviewer: Indentation is off + need spacing here as well

```
### Type params of the transport
`HttpTransport` accepts one type param that describes the request and response types for the endpoint and the transport. It is used to ensure that the transport and endpoint are compatible.
```typescript
const transport = new HttpTransport<EndpointTypes>({
```
Example structure of `EndpointTypes`

> ```typescript

Reviewer: Should all be one block?

export type EndpointTypes = {  
 Request: {  

> 	Params: CryptoRequestParams // type of input parameters

Reviewer: ```suggestion
	Params: CryptoRequestParams // type for the EA endpoint input parameters
```

  }  

>  Response: SingleNumberResultResponse // type of External Adapte response. `SingleNumberResultResponse` is built in type that indicates that both `data` and `result` are numbers

Reviewer: ```suggestion
 Response: SingleNumberResultResponse // type of External Adapter response. `SingleNumberResultResponse` is a built in type that indicates that both `data` and `result` are numbers
```

 Settings: BaseAdapterSettings  // type of adapter config
 Provider: {  

>   RequestBody: never  // type of request body in case of POST requests. `never` for GET requests

Reviewer: You should technically be able to have a request body in a GET request, but usually not. Just a nit pick

```suggestion
  RequestBody: never  // type of request body in the case of POST requests. This is usually `never` for GET requests, and is *not* the same as query params
```

      response: {
        data: {
          result: value,
        },
        result: value,
      },
    }
  })
}
```

### Type params of the transport

`HttpTransport` accepts one type param that describes the request and response types for the endpoint and the transport. It is used to ensure that the transport and endpoint are compatible.

```typescript
const transport = new HttpTransport<EndpointTypes>({
```

Example structure of `EndpointTypes`

```typescript
export type EndpointTypes = {  
 Request: {  
	Params: CryptoRequestParams // type for the EA endpoint input parameters
  }  
 Response: SingleNumberResultResponse // type of External Adapter response. `SingleNumberResultResponse` is a built in type that indicates that both `data` and `result` are numbers
 Settings: BaseAdapterSettings  // type of adapter config
 Provider: { 
  RequestBody: never  // type of request body in the case of POST requests. This is usually `never` for GET requests, and is *not* the same as query params
  ResponseBody: ProviderResponseBody  // interface of raw response body from Data Provider 
  }  

> }

Reviewer: Nitpick, but we're mixing tabs/spaces here, and there's some trailing whitespace on some lines

```

> 
>  

Reviewer: ```suggestion

```
